### ISSUE TITLE
"File locked" failure during upload

### ISSUE THREAD

**Issue by stevejalim** [mozilla/wagtail-localize-smartling#45]
I submitted five pages (therefore five Jobs) in one sync and couple of them failed. One failure was related to a failure generating a visual context, but the other was a ValidationError returned by Smartling:

`2024-12-03 14:20:09.478: {'response': {'code': 'VALIDATION_ERROR', 'errors': [{'key': None, 'message': 'File locked: job_7.po', 'details': None}]}}`

This is hard to replicate locally, but causes a fatal error that leaves us in a tricky state
```[tasklist]
### Tasks
- [x] Diagnose root cause
- [ ] Fix
- [ ] Look at ways to recover gracefully - either auto-deleting the job in Smartling or at least logging that that's what needs to happen manually
```

**Comment by zerolab** [comment]
https://help.smartling.com/hc/en-us/articles/1260805481410-Error-Codes has

<p><strong>423 Locked</strong></p>

Response Code | Possible Causes
-- | --
RESOURCE_LOCKED | The requested file is currently being processed by another operation. The file will be unlocked after that operation completes.

it is possible there is some process on the Smartling side between translations submitted and PO file ready to be downloaded

**Comment by stevejalim** [comment]
I opened a ticket with Smartling as there was no reference to "File locked" in their docs. They responded with

> The 'File Locked' response indicates that the file was still processing when your API integration tried to add it to the Job. The file will be unlocked after the processing operation completes.
 
> In order to address this, it is recommended that you use the Job Batches V2 API to add files to the Job. The correct usage of the Jobs Batches API is to [Create a Job](https://api-reference.smartling.com/#tag/Jobs/operation/addJob) using the Jobs API, [Create a Batch](https://api-reference.smartling.com/#tag/Job-Batches-V2/operation/createJobBatchV2) specifying your file(s) upfront, then Upload your File(s) to the Batch [Upload File to Batch](https://api-reference.smartling.com/#tag/Job-Batches-V2/operation/uploadFileToJobBatchV2). The files then get processed & added to the Job without you needing another call to do so, eliminating the asynchronous issues of processing.

![unnamed](https://github.com/user-attachments/assets/ad29f7eb-1cf5-4c78-ae75-ecb4bf9bed14)

> As such, if your integration added files to the Batch using Job Batches API instead of adding files to the Job using the Jobs API, this should ensure that you do not encounter 423 (File Locked) responses. 
 
> For more details, you may wish to refer to the following section of our Help Center: https://help.smartling.com/hc/en-us/articles/1260805481410-Error-Codes and API docs: https://api-reference.smartling.com/#tag/Job-Batches-V2

**Comment by stevejalim** [comment]
So, taking in the above and looking at the code:

https://github.com/mozilla/wagtail-localize-smartling/blob/bbb38787e83ab5768568f2b84d59e4014ffe1f16/src/wagtail_localize_smartling/sync.py#L105-L111

...I'm assuming that the blow up happens because at the moment we're uploading a file and then attaching it to a job before Smartling has had time to process it. 

If that's right, we could

1) Sleep a little and poll Smartling about the [state of the upload](https://api-reference.smartling.com/#tag/Files/operation/getFileTranslationStatusAllLocales) until we know we can continue safely

2) Use the async batches approach recommended by Smartling support - even though we're only uploading one file at a time, the batches approach has a chain-reaction effect that should avoid the locked file issue.

I'm going with Option 2, but if for some reason that's not viable, then Option 1 doesn't look too bad

**Comment by stevejalim** [comment]
> https://help.smartling.com/hc/en-us/articles/1260805481410-Error-Codes has
> 
> **423 Locked**
> Response Code 	Possible Causes
> RESOURCE_LOCKED 	The requested file is currently being processed by another operation. The file will be unlocked after that operation completes.
> 
> it is possible there is some process on the Smartling side between translations submitted and PO file ready to be downloaded

In the case of the error, it was during job creation/upload, rather than download. I think the explanation from Smartling support covers what may have been up, though

**Comment by stevejalim** [comment]
Hmm - to add a Batch, apparently we need to specify `workflowLocales` -- which doesn't appear to be exposed by the API...

**Comment by stevejalim** [comment]
OK, this is working fine locally against a sandbox account (but the production API), with 10+ Jobs all being synced at once without issue. Will get https://github.com/mozilla/wagtail-localize-smartling/pull/46 merged and try it out live

**Comment by zerolab** [comment]
Fixed in #45

**Comment by stevejalim** [comment]
Yeap, uploaded a bunch of jobs and things look OK right now
